Added host-side associative scan function

bohnstingl opened the pull request:
This PR aims to implement a host-side associative scan functionality. This functionality does solely rely on PyTorch compiled code and does not explicitly invoke `tl.associative_scan` yet.

This is related to https://github.com/pytorch/pytorch/issues/95408

@Chillee 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

bohnstingl committed:
Added host-side associative scan function

vadimkantorov commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -79,32 +109,150 @@ def add(x: torch.Tensor, y: torch.Tensor):
+    stacked = torch.stack([a, b], dim=dim + 1)
+    interleaved = torch.flatten(stacked, start_dim=dim, end_dim=dim + 1)
+    if b_trunc:
+        # TODO: find torch alternative for slice_along dim for torch.jit.script to work
```
Regarding `slice_along_dim`, maybe this feature request is related: https://github.com/pytorch/pytorch/issues/41625

bohnstingl replied:
Thanks for the pointer! I tried to follow the issues, but is there anything at the moment that I should incorporate?

vadimkantorov replied:
I'm not a maintainer/reviewer. I also missed a more capable/simple slicing function to not have to juggle slice(None) and calculate the correct dimension index, that's all :)

lezcano reviewed (commented):
without proper parallelisation this is going to be very slow. Can you post benchmarks?

bohnstingl commented:
Thank you for your comment. I tried to carry out two benchmarks, one for the case of `cumsum` and one for `mm`. I compared to a naive implementation, with `torch.cumsum` and with JAX. The `cumsum` code that I've used is
```
import torch
from torch._higher_order_ops.associative_scan import associative_scan as device_scan
import torch._dynamo.config as config
import jax
import jax.numpy as jnp
from triton.testing import do_bench
import matplotlib.pyplot as plt
import numpy as np

config.cache_size_limit = 10000
torch.set_default_device('cuda')

def associative_scan_fct(operator, input, dim=0, reverse=False):
    inp_leaves, spec = pytree.tree_flatten(input)
    result_flat = []
    num_leaves = len(inp_leaves)
    op = reversed if reverse else lambda x: x

    for ind in op(range(inp_leaves[0].size(dim))):
        r = [
            inp_leaves[leave_ind][(slice(None),) * dim + (ind,)]
            for leave_ind in range(num_leaves)
        ]
        if (ind > 0 and not reverse) or (
            ind < (inp_leaves[0].size(dim) - 1) and reverse
        ):
            r = operator(
                pytree.tree_unflatten(result_flat[-1], spec),
                pytree.tree_unflatten(r, spec),
            )
        r_flat, _ = pytree.tree_flatten(r)
        result_flat.append(r_flat)
    results = [
        torch.stack([e[leave_ind] for e in op(result_flat)], dim)
        for leave_ind in range(num_leaves)
    ]
    return pytree.tree_unflatten(results, spec)

comp_host_scan = torch.compile(device_scan)
comp_cumsum = torch.compile(torch.cumsum)

jax_scan = jax.lax.associative_scan
compiled_jax_scan = jax.jit(jax_scan, static_argnums=(0,))

def scan_fct(i, j):
    return i + j

mm_inps = lambda S, B, construct: construct(S, B, B)
input_func = mm_inps

jax_construct = lambda *shape: jnp.ones(shape, dtype=jnp.float32)
torch_construct = lambda *shape: torch.ones(*shape, dtype=torch.float32)
B = 100
jax_times_compiled = []
eager_times = []
compiled_host_side_times = []
compiled_device_side_times = []
upper_limit = 15000
lower_limit = 2000
for S in range(lower_limit, upper_limit, 2000):
    inp = input_func(S, B, torch_construct)
    jax_inp = input_func(S, B, jax_construct)

    eager_times.append(do_bench(lambda: associative_scan_fct(scan_fct, inp, dim=0)))
    compiled_host_side_times.append(do_bench(lambda: comp_host_scan(scan_fct, inp, dim=0, reverse=False, host_side=True)))
    compiled_device_side_times.append(do_bench(lambda: comp_cumsum(inp, 0)))
    jax_times_compiled.append(do_bench(lambda: compiled_jax_scan(scan_fct, jax_inp)))

print("eager: ", eager_times)
print("compiled host-side: ", compiled_host_side_times)
print("compiled device-side: ", compiled_device_side_times)
print("jax compiled: ", jax_times_compiled)

plt.figure()
# plt.plot(np.arange(lower_limit, upper_limit, 2000), eager_times, label='Eager')
plt.plot(np.arange(lower_limit, upper_limit, 2000), compiled_host_side_times, label='Associative scan host-side compiled')
plt.plot(np.arange(lower_limit, upper_limit, 2000), compiled_device_side_times, label='Associative scan device-side compiled')
plt.plot(np.arange(lower_limit, upper_limit, 2000), jax_times_compiled, label='Jax compiled')
plt.xlabel('Length')
plt.ylabel('Time [s]')
plt.legend()
plt.show()
```
<img width="551" alt="image" src="https://github.com/pytorch/pytorch/assets/2130441/a298e818-b966-4d1f-afc3-3c58955c56eb">
I acknowledge that the host-side scan is slower compared to JAX or torch.cumsum in this scenario, but it is way faster compared to an eager implementation in pure PyTorch (which I measured, but didn't even plot).

For the `mm` scenario, the numbers look even better, with the host-side scan being slightly faster than JAX.
```
import torch
from torch._higher_order_ops.associative_scan import associative_scan as device_scan
import torch._dynamo.config as config
import jax
import jax.numpy as jnp
from triton.testing import do_bench
import matplotlib.pyplot as plt
import numpy as np

config.cache_size_limit = 10000
torch.set_default_device('cuda')

def associative_scan_fct(operator, input, dim=0, reverse=False):
    inp_leaves, spec = pytree.tree_flatten(input)
    result_flat = []
    num_leaves = len(inp_leaves)
    op = reversed if reverse else lambda x: x

    for ind in op(range(inp_leaves[0].size(dim))):
        r = [
            inp_leaves[leave_ind][(slice(None),) * dim + (ind,)]
            for leave_ind in range(num_leaves)
        ]
        if (ind > 0 and not reverse) or (
            ind < (inp_leaves[0].size(dim) - 1) and reverse
        ):
            r = operator(
                pytree.tree_unflatten(result_flat[-1], spec),
                pytree.tree_unflatten(r, spec),
            )
        r_flat, _ = pytree.tree_flatten(r)
        result_flat.append(r_flat)
    results = [
        torch.stack([e[leave_ind] for e in op(result_flat)], dim)
        for leave_ind in range(num_leaves)
    ]
    return pytree.tree_unflatten(results, spec)

comp_device_scan = torch.compile(device_scan)

jax_scan = jax.lax.associative_scan
compiled_jax_scan = jax.jit(jax_scan, static_argnums=(0,))

def mm_chain_scan_jax(i, j):
    return jax.vmap(jnp.matmul)(i, j)

def mm_chain_scan_torch(i, j):
    return torch.vmap(torch.mm)(i, j)

def mm_chain_scan_torch_wovmap(i, j):
    return torch.mm(i, j)

mm_inps = lambda S, B, construct: construct(S, B, B)
scan_func = mm_chain_scan_torch
scan_func_wovmap = mm_chain_scan_torch_wovmap
scan_func_jax = mm_chain_scan_jax
input_func = mm_inps

jax_construct = lambda *shape: jnp.ones(shape, dtype=jnp.float32)
torch_construct = lambda *shape: torch.ones(*shape, dtype=torch.float32)
B = 100
jax_times_compiled = []
eager_times = []
compiled_host_side_times = []
upper_limit = 15000
lower_limit = 2000
for S in range(lower_limit, upper_limit, 2000):
    inp = input_func(S, B, torch_construct)
    jax_inp = input_func(S, B, jax_construct)

    eager_times.append(do_bench(lambda: associative_scan_fct(scan_func_wovmap, inp, dim=0)))
    compiled_host_side_times.append(do_bench(lambda: comp_device_scan(scan_func, inp, dim=0, reverse=False, host_side=True)))
    jax_times_compiled.append(do_bench(lambda: compiled_jax_scan(scan_func_jax, jax_inp)))

print("eager: ", eager_times)
print("compiled host-side: ", compiled_host_side_times)
print("jax compiled: ", jax_times_compiled)

plt.figure()
plt.plot(np.arange(lower_limit, upper_limit, 2000), compiled_host_side_times, label='Associative scan host-side compiled')
plt.plot(np.arange(lower_limit, upper_limit, 2000), jax_times_compiled, label='Jax compiled')
plt.legend()
plt.show()
```
<img width="563" alt="image" src="https://github.com/pytorch/pytorch/assets/2130441/1851d9a7-0b4b-44da-b178-7a7ae40b8446">

As a side note, this code is largely based on the code from @Chillee.

lezcano commented:
Yeah, I don't have that much of an issue with the scaling with the batch size. I do have it with the scan dimension, which is where the work is not done in parallel.

Chillee commented:
@lezcano This *is* parallelized along the scan dimension - the scan elements are batched together using a parallel scan algorithm.

bohnstingl commented:
Is there any way that one can introduce more parallelism or any other way to make it more efficient?

lezcano commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -79,32 +109,150 @@ def add(x: torch.Tensor, y: torch.Tensor):
+                # return associative_scan_host_side(combine_fn, input, dim, False)
+            else:
+                if reverse:
+                    raise ValueError(
+                        "For the device-side associative scan, the reverse flag has not yet been implemented"
+                    )
```
We could do two flips here as well to support it.

bohnstingl replied:
I did this in my clean-up as well. One might though need to change the interface of `torch.cumsum` and `torch.cumprod`  to include that parameter, I guess?

lezcano commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -40,10 +41,39 @@ def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
     input: pytree.PyTree,
     dim: int,
+    reverse: bool = False,
+    host_side: bool = False,
```
why do we want to have a flag, rather than dispatching using the devices in the input. We could return a function that does
```python
def associative_scan(input, ...):
  if input.is_cpu: ...
  else: ...
```
and have `torch.compile`  trace appropriately.

bohnstingl replied:
I think that there might be a confusion here. The _host-side_ here does not refer to CPU-only in this case. Rather it refers to the fact that no _device-side_ operations, like the `tl.associative_scan` are invoked. I added a testcase to check for the compatibility for CUDA as well. 

Interestingly to me though, the CUDA test fails for me, not because of the device, but rather because when after using the `reverse=False` flag and compiling the function, the reverse is set to `True` and invoked again, it doesn't issue a new compilation and hence still uses the `reverse=False `during execution, which then fails the test. This only happens on the CUDA device, not on the CPU. Do you have any suggestions as to how I can resolve this?

lezcano replied:
Oh, I see. Then it'd be nice to bikeshed the name of the parameter a bit...

Can you give a small repro for the issue?

bohnstingl replied:
What about if we reverse the meaning and have a parameter like `use_explicit_tl` that would default to `True` to not change the original behavior?

Of course. This code snippet can reproduce the error that I am seeing. If one sets the device to `torch.device('cpu')` it will work, if one sets the device to `torch.device('cuda')`, it will error out.
```
def fct(x: torch.Tensor, y: torch.Tensor):
    return x

x = torch.randn(3, 2, 2, requires_grad=True, device=torch.device('cuda'))
result = associative_scan(fct, x, 0, host_side=True, reverse=False)
expexted_result = _fake_associative_scan(fct, x, 0, reverse=False)
self.assertEqual(result, expexted_result)
result = associative_scan(fct, x, 0, host_side=True, reverse=True)
expexted_result = _fake_associative_scan(fct, x, 0, reverse=True)
self.assertEqual(result, expexted_result)
```

lezcano replied:
For the keyword, what about something like `generic_scan=False` as the default, and writing a good documentation explaining the pros and cons (capabilities vs. performance) of using this flag?

About that issue... I don't know off the top of my head. The only thing that comes to mind is that somehow `torch.compile` may not be caching that flag and that's why it's not recompiling it? But I don't see a priori why would that be the case.

bohnstingl replied:
Right, to me this sounds good. I have updated the code accordingly.

Regarding this issue, I was wrong in my initial assessment. After digging into this issue, I found that the function behaves differently when using `torch.compile` or not. For example, this code
```
def fct(x: torch.Tensor, y: torch.Tensor):
    return x + y

for n in range(15):
    x = torch.arange(n, device=torch.device('cuda'))
    associative_scan1 = torch.compile(associative_scan)
    associative_scan2 = associative_scan
    result1 = associative_scan1(fct, x, 0, True, True)
    result2 = associative_scan2(fct, x, 0, True, True)
    print(str(n), ': ', ['{:2d}'.format(int(r.cpu().numpy())) for r in result1])
    print(str(n), ': ', ['{:2d}'.format(int(r.cpu().numpy())) for r in result2])
    print(str(n), ': ', [' T' if result1[i] == result2[i] else ' F' for i in range(n)])
    print('-'*80)
```
yields 
```
0 :  []
0 :  []
0 :  []
--------------------------------------------------------------------------------
1 :  [' 0']
1 :  [' 0']
1 :  [' T']
--------------------------------------------------------------------------------
2 :  [' 1', ' 1']
2 :  [' 1', ' 1']
2 :  [' T', ' T']
--------------------------------------------------------------------------------
3 :  [' 3', ' 3', ' 2']
3 :  [' 3', ' 3', ' 2']
3 :  [' T', ' T', ' T']
--------------------------------------------------------------------------------
4 :  [' 6', ' 6', ' 6', ' 3']
4 :  [' 6', ' 6', ' 5', ' 3']
4 :  [' T', ' T', ' F', ' T']
--------------------------------------------------------------------------------
5 :  ['10', '10', ' 9', ' 7', ' 4']
5 :  ['10', '10', ' 9', ' 7', ' 4']
5 :  [' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
6 :  ['15', '15', '14', '12', ' 9', ' 5']
6 :  ['15', '15', '14', '12', ' 9', ' 5']
6 :  [' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
7 :  [' 0', ' 0', '21', '21', '18', '18', ' 6']
7 :  ['21', '21', '20', '18', '15', '11', ' 6']
7 :  [' F', ' F', ' F', ' F', ' F', ' F', ' T']
--------------------------------------------------------------------------------
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8] torch._dynamo hit config.cache_size_limit (8)
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8]    function: 'inner' (/data_malta3_ssd/pytorch_git/torch/_dynamo/external_utils.py:36)
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8]    last reason: tensor 'L['args'][1]' size mismatch at index 0. expected 0, actual 8
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
W0625 20:25:33.179000 139648747538240 torch/_dynamo/convert_frame.py:762] [0/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
8 :  ['28', '28', '27', '25', '22', '18', '13', ' 7']
8 :  ['28', '28', '27', '25', '22', '18', '13', ' 7']
8 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
9 :  ['36', '36', '35', '33', '30', '26', '21', '15', ' 8']
9 :  ['36', '36', '35', '33', '30', '26', '21', '15', ' 8']
9 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
10 :  ['45', '45', '44', '42', '39', '35', '30', '24', '17', ' 9']
10 :  ['45', '45', '44', '42', '39', '35', '30', '24', '17', ' 9']
10 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
11 :  ['55', '55', '54', '52', '49', '45', '40', '34', '27', '19', '10']
11 :  ['55', '55', '54', '52', '49', '45', '40', '34', '27', '19', '10']
11 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
12 :  ['66', '66', '65', '63', '60', '56', '51', '45', '38', '30', '21', '11']
12 :  ['66', '66', '65', '63', '60', '56', '51', '45', '38', '30', '21', '11']
12 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
13 :  ['78', '78', '77', '75', '72', '68', '63', '57', '50', '42', '33', '23', '12']
13 :  ['78', '78', '77', '75', '72', '68', '63', '57', '50', '42', '33', '23', '12']
13 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
--------------------------------------------------------------------------------
14 :  ['91', '91', '90', '88', '85', '81', '76', '70', '63', '55', '46', '36', '25', '13']
14 :  ['91', '91', '90', '88', '85', '81', '76', '70', '63', '55', '46', '36', '25', '13']
14 :  [' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T', ' T']
-------------------------------------------------------------------------------- 
```
I tried to track it down further, but unfortunately I have no clue where this could originate from.

lezcano commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -79,32 +109,150 @@ def add(x: torch.Tensor, y: torch.Tensor):
 
-    result_flat = associative_scan_op(combine_fn, leaves, dim)
+    if host_side:
+        with _set_compilation_env():
+            with torch._dynamo.utils.disable_cache_limit():
+                with _temp_remove_pre_dispatch_torch_function_mode():
+                    return torch.compile(
```
why is this necessary?

bohnstingl replied:
I removed it during the clean-up.

lezcano reviewed (commented):
Ah, I see in which sense this is parallel. Given how different it is to the algorithm used on GPU, it would still be good to benchmark the operation varying the batched dimension to see what performance characteristics it has.

Overall the idea looks good, but I left a few comments.

lezcano commented:
Also, I see no reason why this algorithm should just be used on CPU. We could also use it to support more generic scans on device, as triton just supports pointwise ops.

vadimkantorov commented:
For general scans:
- it would be good (even if inefficient) supporting arbitrary cells (e.g. passing a custom RNN cell to the scan primitive and getting a somewhat perf/memory efficient custom RNN model)
- it might be good to support loop unrolling during inference or CUDA codegen: e.g. compile together several iterations of the loop to minimize number of kernel launches (maybe this could be some sort of torch.compile arg) for long sequences, in this case we could have number of kernels reduced to e.g. SEQLEN/8, and for sequences the per-step kernel launch overhead can be signficant

bohnstingl committed:
Integrated comments from first review round + clean-up

bohnstingl commented:
@lezcano Thank you very much for your comments. Indeed, this feature is not intended to be for CPU only. The name _host-side_ might be misleading here. The feature performs the associative scan operation on the _host-side_ in the sense that it does not utilize the triton, e.g. _tl.associative_scan_, or _device-specific_ features. The feature can be used with GPUs and with CPUs. I a separate testcase for that as well.

@vadimkantorov This feature partly fills this gap if the RNN that you are feeding is linear, e.g., it can be performed in an associative way. The feature for an arbitrary RNN is certainly something that I am also highly interested, but the closest that feature that I can see in this direction is a [while_loop](https://github.com/pytorch/pytorch/pull/124573) operator, where the body is your arbitrary RNN.

Your second idea sounds really interesting, but I wouldn't really know how to do this. Could you help me out with this?

vadimkantorov commented:
I think both ideas can be in the beginning only tests/benchmarks.

Regarding on how to unroll the loop and when we can compile several loop iterations in a kernel - I don't know enough on how to implement this in torch.compile/Inductor. But I heard at some conference that someone used this trick to accelerate WaveRNN inference or sth like that. I guess memory placement optimization can be important here too and maybe for small enough RNN states (like 32-dim or 64-dim or 128-dim like in https://github.com/NVlabs/tiny-cuda-nn) some extra memory optimizations can be done

bohnstingl committed:
Changed naming of parameter to general_scan

lezcano commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -63,8 +92,15 @@ def associative_scan(
+        dim (int): The dimension to scan over
+        reverse (bool): A boolean stating if the scan should be reversed with respect to the dimension.
+        generic_scan (bool): A boolean stating whether a generic scan mode should be used. 
+            If the generic scan mode is ``not used``, there are restrictions on the operations allowed
+            within the ``combine_fn``. For example, only pointwise functions are currently allowed.
+            If the generic scan mode is ``used``, the restrictions on the combine_fn are less strict, 
+            for example also non-pointwise operations are allowed. 
+            However, the non-generic scan mode utilizes efficient ``tl.associative_scan`` calls and is thus
+            more efficient as the generic scan mode.
```
A suggestion for how to write it a bit more concisely.
```suggestion
        If ``generic_scan=False``, only pointwise ops may be used within ``combine_op``. ``generic_scan=False``
        may just be used on CUDA tensors. On the other hand, ``generic_scan=False`` should be more
        efficient than ``generic_scan=True`` whenever it can be used.
```

bohnstingl replied:
Done

lezcano commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -63,8 +92,15 @@ def associative_scan(
+            However, the non-generic scan mode utilizes efficient ``tl.associative_scan`` calls and is thus
+            more efficient as the generic scan mode.
 
     Example::
```
Can you add an example of `generic_scan=True` (one with matmul would be good) and one with two inputs and outputs (e.g. something like `cummin`).
In general, I would like to show that the API for `generic_scan=True` and `=False` is different (!!) when there are multiple inputs and outputs.

bohnstingl replied:
I tried to unify the API and hence this may not be necessary anymore?

lezcano commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -63,8 +92,15 @@ def associative_scan(
             This function must be pure, pointwise, and satisfy the associative property.
```
The comment in line 86 above is stale, can you update it?

bohnstingl replied:
Done

lezcano commented on test/functorch/test_control_flow.py:
```diff
@@ -431,6 +462,260 @@ def fwbw(map_op, f, x, y):
+
+        for direction in [False, True]:
+            result = associative_scan(add, x, 0, generic_scan=True, reverse=direction)
+            expexted_result = _fake_associative_scan(add, x, 0, reverse=direction)
```
nit. expected
There are a few expexted in the test

bohnstingl replied:
Done

lezcano reviewed (commented):
It'd be nice if we could use this mode automatically as a fallback, but we currently can't, because in this mode we have a different API when we have multiple inputs-outputs. In the `generic_scan=False` case we need to unpack the inputs and outputs (`combine_op` takes 2n inputs and returns n outputs) while in the `generic_scan=True` the inputs are packed.
Do you reckon it'd be possible to reconcile both APIs? If so, it'd be great to remove the flag, and use `generic_scan=False` if the inputs are in cuda and the combine op is all pointwise ops, and otherwise use the other one as a fallback.

bohnstingl committed:
Unified API between generic_scan and non generic_scan
Fixed typos
Updated documentation

bohnstingl commented:
I tried to unify the `generic_scan=False` and `generic_scan=True` case. The default is `generic_scan=False`, but there are now checks internally to determine whether the `generic_scan` has to be used. Specifically `generic_scan=True` if:
1. Not all tensors are on CUDA
2. The operations inside the graph of the function is used by a non-pointwise function
However, I thought that it might be still good that the user can `force` the `generic_scan=True` behavior, so the user can override `generic_scan` to be true.

@lezcano please let me know whether the checks [here](https://github.com/pytorch/pytorch/pull/129307/commits/106f90673c05fcfb504bf5ae1fde8adb95f0aff0#diff-43442eb487890d221156357e289c9ba39d7ee3da0a96bd7b13448b4716a4cfcaR71-R84) are appropriate.

Furthermore, I spotted an issue when the device `CUDA` is used. I tried to dig into this a little bit [here](https://github.com/pytorch/pytorch/pull/129307#discussion_r1653340734) and it appears that for some lengths of the input tensors, the results are sporadically wrong and I don't know why. I also added TODOs in the [testcases](https://github.com/pytorch/pytorch/pull/129307/commits/106f90673c05fcfb504bf5ae1fde8adb95f0aff0#diff-046f76d8f6c9786a877e46b9e686efd7a2a60a229d43cea3bc40b7461cf2b998R549-R551) that can reproduce this behavior. Could you help me out with this one?

bohnstingl committed:
Moved check of arguments

lezcano reviewed (commented):
So, I think the current code that looks into whether a function is pointwise may be not general enough. At the moment, these checks are done in `lower_pointwise_subgraph`, within the lowering of this OP. The ideas I have to mix both approaches are not super clean.
Pinging @peterbell10 see if he has any ideas on what's the best approach here, and for a general review / help with those issues you are seeing.

On a different note, if we are able to choose the algorithm cleanly, I would propose to not expose the flag now in the interest of "it's easy to add a flag later if people request it, but it's very difficult to deprecate a given flag".

Also, I assume that this is BC breaking to some extent, as it changes the way `associative_scan` works for multiple inputs? I think it's fine to do this, as we are not even rendering the docs for this OP (which we should fix at some point... not this PR tho)

vadimkantorov commented:
If the auto check for pointwiseness is too complex, better to actually intoduce it as a separate function (instead of a flag)?

Chillee commented:
I think we should explicitly expose a flag for whether it's device-side or host-side. Even if it's all pointwise ops users may want to try the host-side scan regardless.

bohnstingl commented:
Yeah, so that's what I initially thought. The user can use this flag to 'force' the `generic_scan`. The other way around may not make so much sense, as the `generic_scan=False` has limitations.

peterbell10 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -40,10 +42,55 @@ def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
+    ), "The pytree of the output of the operator needs to match the input pytree"
+    
+    generic_scan_required = False
+    combine_fn = make_fx(combine_fn)(pytree.tree_unflatten(
```
Do we have to run this check under dynamo? I would have thought we could call `associative_scan_op`, and inside the compiler perform this check and if it fails we use the eager implementation as a decomposition.

bohnstingl replied:
Thank you for your review and your comment. Unfortunately, I don't fully understand it and I was wondering whether you could elaborate a bit more on it for me?
I was under the impression that the goal is to have a `generic_scan` (which can either run in `eager` mode or under `torch.compile`) that is more general than the optimized associative scan that invokes `tl.associative_scan`. The idea was then to determine apriori which version of scan to use, the generic scan or the non generic scan. This involves checking whether the `combine_fn` contains only pointwise operations and I didn't know about any better way to do this, but this check fails under dynamo. I am not sure how and where you envision to move this check?

Also, regarding your comment below. I just used a similar methodology as in other HOPs, such as [cond](https://github.com/pytorch/pytorch/blob/998e72cd435e73700d84aa489277866e2d12bc59/torch/_higher_order_ops/cond.py#L139-L144). I thought that the result should be that the `associative_scan` function is always compiled and that's why this if clause is placed there. Are you suggesting to remove this and expect the user to invoke `torch.compile` instead, because there could be cases where one doesn't want to have this function compiled?

peterbell10 replied:
> I am not sure how and where you envision to move this check?

Inside the compiler when compiling `associative_scan_op` we can choose to fallback to the generic implementation instead. Ideally, this could be done in `torch/_inductor/decompositions.py` .

> Also, regarding your comment below. I just used a similar methodology as in other HOPs, such as [cond](https://github.com/pytorch/pytorch/blob/998e72cd435e73700d84aa489277866e2d12bc59/torch/_higher_order_ops/cond.py#L139-L144)

Looks like cond is being compiled with the eager backend there which doesn't make a great deal of sense to me tbh. 

cc @zou3519 @aakhundov  any comment on why HOPs are being compiled with eager backend instead of just running as normal eager code?

bohnstingl replied:
cc @ydwu4

ydwu4 replied:
> any comment on why HOPs are being compiled with eager backend instead of just running as normal eager code

We rely on dynamo to 1. check and raise error when there're any side-effects in the udf 2. lift closures as implicit inputs 3. flatten/unflatten pytree.

peterbell10 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -79,32 +132,116 @@ def add(x: torch.Tensor, y: torch.Tensor):
-                combine_fn, input, dim
-            )
-
+            return torch.compile(associative_scan, fullgraph=False)(
```
If we have an eager implementation then we shouldn't need to compile this.

bohnstingl committed:
Temporarily removed check of combine_fn for pointwise operations

bohnstingl commented:
@peterbell10 
I was wondering whether you could help me out on two issues that this feature currently is stuck on
1. Issue with the “random” results that I observe when using the `CUDA` device: There is a [testcase](https://github.com/pytorch/pytorch/blob/36180ac20a5e542ea192070c088257db7c122800/test/functorch/test_control_flow.py#L465-478) that shows that for some input length the results between the compiled and the non-compiled version differ. This is only the case though for the `CUDA` device. For the `CPU` device, both behaviors are the same. You can find my current investigations on this in the thread [above](https://github.com/pytorch/pytorch/pull/129307#discussion_r1653340734)
2. Issue with the [checks](https://github.com/pytorch/pytorch/blob/36180ac20a5e542ea192070c088257db7c122800/torch/_higher_order_ops/associative_scan.py#L72-L81) that the combine_fn only contains pointwise operations. You mentioned that you wanted to do the compilation or the checks in a different form, but I don't quite know what you had in mind. Could you maybe give me some guidance?

Thank you in advance

bohnstingl committed:
Integrated generic_scan as eager function
Integrated generic_scan as AutoGrad function
Added additional testcases

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Added additional testcases

bohnstingl commented:
I have cleaned-up the code, did a rebase to main and integrated the `generic_scan` as the eager function as well as for the AutoGrad.

Currently there are two issues open:

1. The fallback to the eager function while compiling `associative_scan`. I am unsure how to do this in an appropriate manner. I initially tried to integrate the checks into torch/_dynamo/variables/higher_order_ops.py, but with little success. Mainly because once the graph is compiled, the ops of the nodes are mostly call_function, for which the `is_pointwise_use` check that I used so far fails.
I looked at `torch/_inductor/decompositions.py` that @peterbell10 mentioned, but to me this seems to be at a too low level, e.g., the functions inside `torch/_inductor/decompositions.py` are invoked from many other places and I don’t know how or where I should place the checks for the `associative_scan`?
Furthermore, I don’t know how one can "break" the compilation in a way that the "eager mode generic scan" gets compiled instead. Can anybody maybe point me to any code snippets that aim to do something similar, or can you maybe give me some more guidance?
2. For the case of the `reverse` flag being set to True, there is a strange behavior with CUDA devices (see [here](https://github.com/pytorch/pytorch/pull/129307#discussion_r1653340734)). I created two specific testcases to highlight this further. The [CPU testcase](https://github.com/pytorch/pytorch/blob/c1df0cb0fbbad5168195f46f21753b8380e89c5a/test/functorch/test_control_flow.py#L466) passes without issues, but the [CUDA testcase](https://github.com/pytorch/pytorch/blob/c1df0cb0fbbad5168195f46f21753b8380e89c5a/test/functorch/test_control_flow.py#L525) shows a weird behavior for certain lengths of the input. I tried to track it down further and this erroneous behavior only appears if `associative_scan` is torch.compiled and then reverse=True. If the flip operations are applied outside of associative_scan, this problem does not appear. In the eager mode and for the CPU device, this does also not happen. Does anybody have some idea where this could come from and how this can be resolved?

@lezcano @Chillee @peterbell10

bohnstingl committed:
WIP: Fallback to eager handling

bohnstingl committed:
WIP: Implemented fallback from associative_scan to generic_associative_scan

bohnstingl commented:
I tried to implement the fallback from `associative_scan` to `generic_associative_scan`. @peterbell10 or @Chillee, could you please take a look and tell me what you think?

I found one issue with the current way of checking for a non-pointwise subgraph [here](https://github.com/bohnstingl/pytorch/blob/associative_scan_host/torch/_higher_order_ops/associative_scan.py#L87-L92). When using CUDA and with `fullgraph=True` during torch.compile, an exception is raised. This behavior is tested in this [testcase](https://github.com/bohnstingl/pytorch/blob/c460229c7b9aa336f334ded384e4926469ebff5a/test/functorch/test_control_flow.py#L1127).
```
Traceback (most recent call last):
  File "/data_malta3_ssd/pytorch_git/torch/testing/_internal/common_utils.py", line 2814, in wrapper
    method(*args, **kwargs)
  File "/data_malta3_ssd/pytorch_git/test/functorch/test_control_flow.py", line 1139, in test_generic_associative_scan_compile_GPU_fullgraph
    result_comp = associative_scan_comp(
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/eval_frame.py", line 435, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/convert_frame.py", line 1121, in __call__
    return self._torchdynamo_orig_callable(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/convert_frame.py", line 472, in __call__
    return _compile(
           ^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_utils_internal.py", line 85, in wrapper_function
    return StrobelightCompileTimeProfiler.profile_compile_time(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_strobelight/compile_time_profiler.py", line 129, in profile_compile_time
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/conda/pt22/lib/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/convert_frame.py", line 817, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/utils.py", line 234, in time_wrapper
    r = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/convert_frame.py", line 636, in compile_inner
    out_code = transform_code_object(code, transform)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/bytecode_transformation.py", line 1270, in transform_code_object
    transformations(instructions, code_options)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/convert_frame.py", line 178, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/convert_frame.py", line 582, in transform
    tracer.run()
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2489, in run
    super().run()
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 904, in run
    while self.step():
          ^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 816, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 510, in wrapper
    return inner_fn(self, inst)
           ^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 1534, in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 754, in call_function
    self.push(fn.call_function(self, args, kwargs))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/variables/lazy.py", line 132, in realize_and_forward
    return getattr(self.realize(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/variables/functions.py", line 302, in call_function
    return super().call_function(tx, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/variables/functions.py", line 102, in call_function
    return tx.inline_user_function_return(self, [*self.self_args(), *args], kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 760, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2704, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2820, in inline_call_
    tracer.run()
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 904, in run
    while self.step():
          ^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 816, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 510, in wrapper
    return inner_fn(self, inst)
           ^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2096, in CALL
    self.call_function(fn, args, kwargs)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 754, in call_function
    self.push(fn.call_function(self, args, kwargs))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/variables/functions.py", line 302, in call_function
    return super().call_function(tx, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/variables/functions.py", line 102, in call_function
    return tx.inline_user_function_return(self, [*self.self_args(), *args], kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 760, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2704, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2820, in inline_call_
    tracer.run()
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 904, in run
    while self.step():
          ^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 816, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 510, in wrapper
    return inner_fn(self, inst)
           ^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 2096, in CALL
    self.call_function(fn, args, kwargs)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/symbolic_convert.py", line 754, in call_function
    self.push(fn.call_function(self, args, kwargs))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/variables/functions.py", line 689, in call_function
    unimplemented(msg)
  File "/data_malta3_ssd/pytorch_git/torch/_dynamo/exc.py", line 221, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: 'skip function disable_proxy_modes_tracing in file /data_malta3_ssd/conda/pt22/lib/python3.11/contextlib.py'

from user code:
   File "/data_malta3_ssd/pytorch_git/torch/_dynamo/external_utils.py", line 38, in inner
    return fn(*args, **kwargs)
  File "/data_malta3_ssd/pytorch_git/torch/_higher_order_ops/associative_scan.py", line 180, in associative_scan
    generic_scan_required = check_args(combine_fn, leaves, spec, dim)
  File "/data_malta3_ssd/pytorch_git/torch/_higher_order_ops/associative_scan.py", line 87, in check_args
    with disable_proxy_modes_tracing():

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True


To execute this test, run the following from the base repo dir:
    python test/functorch/test_control_flow.py -k TestControlFlow.test_generic_associative_scan_compile_GPU_fullgraph

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
WIP: Lifting free variables to arguments

bohnstingl committed:
Implemented lifted_arguments

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update (base update)

[ghstack-poisoned]

bohnstingl committed:
Major fixes to testcases and improving functionality of associative_scan

bohnstingl committed:
Merge branch 'gh/peterbell10/773/head' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

peterbell10 committed:
Update

[ghstack-poisoned]

bohnstingl committed:
Merge branch 'gh/peterbell10/773/head' of github.com:pytorch/pytorch into associative_scan_host

peterbell10 committed:
Update

[ghstack-poisoned]

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

peterbell10 committed:
Update

[ghstack-poisoned]

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Small update to testcase

peterbell10 committed:
Update

[ghstack-poisoned]

bohnstingl committed:
Fixed lintrunner issues
Fixed minor issues with testcases
Fixed issues with shape collection for TensorVariables

bohnstingl committed:
Merge branch 'gh/peterbell10/773/head' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Integrated failing flip bad accuracy test to test_repros

bohnstingl committed:
Fixed lint issues
Extended testcase of torch.flip with data types and backends

bohnstingl committed:
Removed unnecessary code

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into associative_scan_host

bohnstingl committed:
Minor update to testcase

bohnstingl committed:
Cosmetic updates

bohnstingl committed:
Fixed vmap functionality of associative_scan
WIP: Issue with view operations in case of matmul

bohnstingl committed:
WIP vmap

ydwu4 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -28,19 +33,98 @@
     input: pytree.PyTree,
     dim: int,
+    reverse: bool = False,
+    generic_scan: bool = False,
```
Should we name them as combine_mode= "pointwise"/"generic"/"auto", where "auto" basically means whenever pointwise fails, fallback to generic. In this pr, I think we should focus adding `pointwise`.

ydwu4 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -71,49 +166,110 @@ def add(x: torch.Tensor, y: torch.Tensor):
 
 def trace_associative_scan(
-    proxy_mode, func_overload, combine_fn: Callable, input: List[torch.Tensor], dim: int
+    proxy_mode,
```
I believe lifted_args should have its separate pr. We should only keep changes related to supporting "generic_associtiave_scan" in this PR. Otherwise it would be difficult to review and track.

ydwu4 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -167,47 +321,59 @@ def associative_scan_op_dense(combine_fn, input, dim):
+def associative_scan_proxy_mode(mode, combine_fn, input, dim, lifted_args):
     if mode.enable_tracing:
-        return trace_associative_scan(mode, associative_scan_op, combine_fn, input, dim)
+        return trace_associative_scan(
```
Can do a rebase? This has changed a bit for other control flow operators.

ydwu4 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -71,49 +166,110 @@ def add(x: torch.Tensor, y: torch.Tensor):
-    assert isinstance(dim, int), "dim must be an int, but got {type(dim)}"
-
-    if not torch._dynamo.is_compiling():
-        with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
```
W'd better still should go through torch.compile for associative_scan to check for any side-effects and lift the free-vars.

ydwu4 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -383,24 +383,22 @@ def _print_ToFloat(self, expr):
-    # torch.arange(n, device="cuda") - 1 and then do a modulus on it
     def _print_PythonMod(self, expr):
-        return " % ".join(map(self.paren, map(self._print, expr.args)))
+        quot, div = expr.args
```
are changes in this file necessary? Can we do it in a saperate pr?

ydwu4 commented on torch/_inductor/runtime/triton_helpers.py:
```diff
@@ -37,6 +37,23 @@ def promote_to_tensor(x):
     return x + tl.zeros((1,), tl.int1)
 
 
+@triton.jit
```
Are these changes necessary?

ydwu4 commented on torch/_inductor/sizevars.py:
```diff
@@ -21,10 +21,11 @@
 from torch.utils._sympy.functions import FloorDiv, ModularIndexing
 from torch.utils._sympy.symbol import symbol_is_type, SymT
-from torch.utils._sympy.value_ranges import bound_sympy
+from torch.utils._sympy.value_ranges import bound_sympy, IntInfinity, ValueRanges
```
What's going on in this file?

ydwu4 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -52,7 +52,7 @@
     import types
 
-    from torch._ops import OpOverload
+    from torch._ops import OperatorBase, OpOverload
```
What's going on here? Should be a separate pr?

ydwu4 commented on torch/onnx/_internal/exporter.py:
```diff
@@ -322,7 +322,7 @@ class ResolvedExportOptions(ExportOptions):
 
     # Private only attributes
-    decomposition_table: dict[torch._ops.OpOverload, Callable]
+    decomposition_table: dict[torch._ops.OperatorBase | torch._ops.OpOverload, Callable]
```
Separate pr?

ydwu4 commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -22,7 +22,7 @@ def __init__(
         diagnostic_context: diagnostics.DiagnosticContext,
         module: torch.fx.GraphModule,
-        decomposition_table: Mapping[torch._ops.OpOverload, Callable],
+        decomposition_table: Mapping[torch._ops.OperatorBase | torch._ops.OpOverload, Callable],
```
Same here.

ydwu4 reviewed (commented):
Went through the diff and left some comments. Can do a rebase? So I could let the CI run.

bohnstingl committed:
Fixed issues with torch.compile(..., dynamic=True)

bohnstingl commented:
Closing this PR.
It has been split into several parts and some of them have landed to `main` already.
Closed:
https://github.com/pytorch/pytorch/pull/133011
https://github.com/pytorch/pytorch/pull/133012

Open:
https://github.com/pytorch/pytorch/pull/134921
https://github.com/pytorch/pytorch/pull/133013
https://github.com/pytorch/pytorch/pull/136966
